### DISCUSSION TITLE
Update Docker tag with multiple major versions

### DISCUSSION THREAD

**Discussion by mbrodala** [renovatebot/renovate#15652] [Request Help]
We use the [`thecodingmachine-php` Docker images](https://github.com/thecodingmachine/docker-images-php).

These use a rather verbose but flexible versioning scheme as can be seen in the README:

`<php-major-minor-version>:<image-version>-<variant>-node<node-version>`

Example: `7.4-v4-fpm-node14`

* PHP version: `7.4`
* Image version: `v4`
* Image variant: `fpm`
* Node version: `14`

(Notice that the Node part is optional, so `<php-major-minor-version>:<image-version>-<variant>` is also valid.)

Renovate already suggests updates for this but as you can imagine it basically suggests updating to the newest version. For the given example this would currently be `8.1-v4-fpm-node14`.

However, it would be very useful if Renovate did suggest the following updates instead:

* `7.4-v4-fpm-node14` -> `8.0-v4-fpm-node14` (PHP: `7.4` -> `8.0`)
* `7.4-v4-fpm-node14` -> `8.1-v4-fpm-node14` (PHP: `7.4` -> `8.1`)
* `7.4-v4-fpm-node14` -> `7.4-v4-fpm-node16` (Node: `14` -> `16`)
* `7.4-v3-fpm-node12` -> `7.4-v4-fpm-node12` (Image: `v3` -> `v4`)

So basically recognize that there are two separate major versions (PHP & Node); these should be updated separately and never at the same time.

I'm rather sure this can be achieved with a custom [`regex` versioning](https://docs.renovatebot.com/modules/versioning/#regular-expression-versioning). But I wonder how I could "fix" some parts here without a huge list of `packageRules` for specific PHP/Node major versions which must be updated manually.

I tried the following but Renovate does not suggest any update here:

```json
  "packageRules": [
    {
      "matchDatasources": ["docker"],
      "matchPackageNames": ["thecodingmachine/php"],
      "versioning": "regex:^\\d+\\.\\d+-v(?<major>\\d+)-\\w(-node\\d+)?$",
      "groupName": "thecodingmachine/php [image]"
    },
    {
      "matchDatasources": ["docker"],
      "matchPackageNames": ["thecodingmachine/php"],
      "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)-v\\d+-\\w(-node\\d+)?$",
      "groupName": "thecodingmachine/php [php]"
    },
    {
      "matchDatasources": ["docker"],
      "matchPackageNames": ["thecodingmachine/php"],
      "versioning": "regex:^\\d+\\.\\d+-v\\d+-\\w-node(?<major>\\d+)$",
      "groupName": "thecodingmachine/php [node]"
    }
  ]
```

Any suggestions?

**Comment by MaronHatoum** [comment]
Hi 
should we open an issue or is there a way to do this using package rules?

**Reply by viceice** [reply]
i don't think this can be easily solved by default, as those tags are not standardized.

so i prefer to add this to the docker docs as sample

**Reply by mbrodala** [reply]
I'm fine with anything as long as it works. ;-)

**Reply by derhuerst** [reply]
Any news on this? I have this problem two with the [`docker.io/postgis/postgis` image tags](https://hub.docker.com/r/postgis/postgis/tags?name=15-3), which combine the PostgreSQL version (e.g. `15`) with the PostGIS version (e.g. `3.5`) and a flavour/compatiblity specifier (e.g. `alpine`).

**Comment by RahulGautamSingh** [comment]
I've looked at this from several angles and there isn't a clean config-level workaround comfortable recommending. (used AI to summarize my findings)

Walking through each attempt with a concrete example using `thecodingmachine/php:7.4-v4-fpm-node14`:                                          
                  
### 1. Versioning regex + package rules can't split one tag into many deps.                                                             
                  
A single tag is one string representing several versions, but Renovate's model is one dependency = one version. 
Versioning regex only changes how versions are compared within a single dep, and package rules only scope behavior, neither can fan one dep out into several PRs.                                                                                                                       
                  
The original attempt from the top of this thread:

```json
  {
    "matchPackageNames": ["thecodingmachine/php"],
    "versioning": "regex:^\\d+\\.\\d+-v(?<major>\\d+)-\\w(-node\\d+)?$"                                                               
  }    
```                                                                                                                               
                                                                                                                                      
This produces no updates at all `\w`matches one char but `fpm` is three, so the regex never matches a real tag. Even after fixing the
regex, multiple package rules on the same dep get merged into a single effective versioning, so you only get one comparison strategy, not three independent updates.       

-----                                                                                     
                  
###   2. Custom regex managers — loose patterns produce silently wrong PRs.

You can extract the tag three times as three virtual deps. If you keep the "other" components loose in the versioning regex:        
  
```json
 "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)-v\\d+-\\w+-node\\d+$"  
```                                                       
                                                                                                                                      
Renovate filters Docker tags to anything matching that shape and compares candidates only by captured PHP `major.minor`. If          `8.1-v3-fpm-node12`, `8.1-v4-fpm-node14`, and `8.1-v5-alpine-node20` all exist, they're all `"PHP 8.1"` to the versioning. Renovate picks one — say `8.1-v5-alpine-node20` — and writes that full tag into the file.                                                            
                  
You asked for a PHP bump and got `7.4-v4-fpm-node14` → `8.1-v5-alpine-node20`: PHP, image rev, variant, and Node all changed.
                                                                                                                                
 ------

### 3. Custom regex managers — hardcoded literals rot.

The fix is to pin the other components as literals so only real, surrounding-matching tags are candidates:                          
  
```json
  // Manager A: bumps PHP only                                                                                                        
  "versioning": "regex:^(?<major>\\d+)\\.(?<minor>\\d+)-v4-fpm-node14$"                                                               
  
  // Manager B: bumps Node only                                                                                                       
  "versioning": "regex:^7\\.4-v4-fpm-node(?<major>\\d+)$"
                                                                                                                                      
  // Manager C: bumps image rev only                                                                                                  
  "versioning": "regex:^7\\.4-v(?<major>\\d+)-fpm-node14$"
```
                                                                                                                                      
Correctness is now fine — Renovate only considers tags with that exact surrounding context. But the moment `Manager B` successfully bumps `Node 14 → 16` in the `Dockerfile`, Manager A's regex still says ...`node14$` and stops matching the current tag entirely. Manager A goes silent until someone hand-edits the 14 to 16. Same for Manager C. Every component bump rots the others.                       
         
-----
         
### 4. Meta-managers don't fix the rot — they multiply it.

The next thought is: add a manager that watches renovate.json itself and updates the stale literals when Node bumps:                
  
```json
  // Manager D (watches renovate.json): rewrites "node14" → "node16" in Managers A and C                                              
  {                                                                                                                                   
    "customType": "regex",
    "fileMatch": ["renovate.json"],                                                                                                   
    "matchStrings": ["node(?<currentValue>\\d+)\\$"],                                                                                 
    "depNameTemplate": "node",
    "datasourceTemplate": "docker",                                                                                                   
    ...                                                                                                                               
  }       
```                                                                                                                            
                                                                                                                                      
But Manager D's own `matchStrings` contains the literal node — and if you want to keep PHP literals in sync too, Manager D's regex    
will reference `7\\.4` as well. So when PHP bumps, Manager D's literals are stale. You'd need Manager G to watch Manager D. Turtles all the way down — the rot just spreads across more files.     

-----                                                                     
                  
### 5. postUpgradeTasks can technically close the rot, but isn't available to most users.                                               
  
The cleanest mechanism is a script that runs in the same PR as the source-file bump and rewrites the stale literals atomically:     
       
```json           
  "postUpgradeTasks": {                                                                                                               
    "commands": [                                                                                                                     
      "scripts/sync-renovate-literals.sh Dockerfile renovate.json"
    ],                                                                                                                                
    "fileFilters": ["renovate.json"]
  }      
```                                                                                                                             
  
The script reads the new Dockerfile tag, parses out each component, and rewrites every pinned regex in `renovate.json` to match. Both file changes land in one commit, no rot.
                                                                                                                                      
Three blockers stop this being a recommended pattern:

  - **Admin-only**. `allowedPostUpgradeCommands` is set in the bot's global config, not repo config. End users on the Mend hosted app or the GitHub App can't enable arbitrary commands.
  - Bespoke per image. `thecodingmachine/php` has four axes in a specific order; `postgis/postgis` has three in a different order; the    
  next compound-tag image is different again. Every user writes and debugs their own script.                                          
  - Subtle interactions with grouping/scheduling. If Manager A and Manager B both have updates in the same run, A's `postUpgradeTask` rewrites B's regex in `renovate.json`, but B already extracted with the old regex earlier in the same run. Easy to get wrong silently.
                  
-----                                                                                                                                 

So every workaround is either incorrect (loose patterns), fragile (hardcoded with rot), multiplicative (meta-managers), or only 
available to self-hosted admins with custom scripting (`postUpgradeTasks`). None of those belong in the docs as a recommended pattern.
                                                                                                                                      
Conclusion is that this needs a first-class feature: a "compound version" dep type where one tag declares multiple independent axes, each with its own datasource and versioning, and Renovate handles reconstruction. 

That's a feature proposal. If anyone wants to attempt that, please open a separate "suggest an idea" discussion and link back here.
